libsupport/include/katana/Result.h

```diff
@@ -265,14 +267,79 @@ ResultErrno() {
 
 namespace internal {
 
+/// Support functions for KATANA_CHECKED
+
+template <class T>
+bool
+CheckedExpressionFailed(const Result<T>& result) {
+  return !result;
+}
+
+inline bool
+CheckedExpressionFailed(const arrow::Status& status) {
+  return !status.ok();
+}
+
+template <class T>
+bool
+CheckedExpressionFailed(const arrow::Result<T>& result) {
+  return CheckedExpressionFailed(result.status());
+}
+
+template <class T>
+ErrorInfo
+CheckedExpressionToError(const Result<T>& result) {
+  return result.error();
+}
+
+inline ErrorCode
+ArrowStatusCodeToKatana(arrow::StatusCode code) {
+  switch (code) {
+  case arrow::StatusCode::Invalid:
+    return ErrorCode::InvalidArgument;
+  case arrow::StatusCode::TypeError:
+    return ErrorCode::TypeError;
+  case arrow::StatusCode::AlreadyExists:
+    return ErrorCode::AlreadyExists;
+  case arrow::StatusCode::KeyError:
+  case arrow::StatusCode::IndexError:
+    return ErrorCode::NotFound;
+  default:
+    return ErrorCode::ArrowError;
+  }
+}
+
+inline ErrorInfo
+CheckedExpressionToError(const arrow::Status& status) {
+  ErrorCode code = ArrowStatusCodeToKatana(status.code());
+  return ErrorInfo(code).WithContext("arrow: {}", status.message());
```

Comment:
I'd personally drop the "arrow" bit because it's not that interesting to general users, but I can be convinced otherwise.